### PR TITLE
Autotune feature

### PR DESCRIPTION
**Pull request by mvdkleijn**
Basic initial feature to allow autotuning of JIRA settings. The initial version supports only JVM min and max memory size tuning.

Tuning types can be set through attribute `['jira']['autotune']['type']` and can contain: 'mixed', 'dedicated' and 'shared'.

* 'mixed' - JIRA and DB run on the same system
* 'dedicated' - JIRA has the system all to itself
* 'shared' - JIRA shares the system with the DB and other applications

Autotune is disabled by default but can be enabled by setting `['jira']['autotune']['enabled']` to boolean true.

Total available memory is auto discovered using Ohai but can be overridden setting your own value in kB using the `['jira']['autotune']['total_memory']` atttribute.

In later stages DB connection parameter tuning will probably be added.

### CODE REVIEW

**Comment by gsreynolds** [comment]
:+1:

File: recipes/autotune.rb [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+# Ohai reports node[:memory][:total] in kB, as in "921756kB"
+mem = total_memory.split("kB")[0].to_i / 1024 # in MB
+
+maximum_memory =
+  { "mixed" => (mem / 100) * 70,
+    "dedicated" => (mem / 100) * 80,
+    "shared" => (mem / 100) * 50
+  }.fetch(tune_type)
+
+node.default['jira']['jvm']['maximum_memory'] = normalize(maximum_memory*1024*1024)
```
Comment (legal90) [inline recipes/autotune.rb]: `normalize` looks really complicated and it's unclear what it returns. How about just append "m" here?
```
node.default['jira']['jvm']['maximum_memory'] = "#{maximum_memory}m"
#...
node.default['jira']['jvm']['minimum_memory'] = "#{minimum_memory}m"
```

File: recipes/autotune.rb [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    raise
+  end
+end
+
+# Parse out total_memory option, or use value detected by Ohai.
+total_memory = node['memory']['total']
+
+if (node['jira'].attribute?('autotune') && node['jira']['autotune'].attribute?('total_memory'))
+  total_memory = node['jira']['autotune']['total_memory']
+  if (total_memory.match(/\A[1-9]\d*kB\Z/) == nil)
```
Comment (legal90) [inline recipes/autotune.rb]: JFYI: [1-9] could be omitted:
```ruby
if !total_memory.match(\A\d+kB\Z)
#...
```

File: recipes/autotune.rb [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    Chef::Application.fatal!([
+        "Bad value (#{total_memory}) for node['jira']['autotune']['total_memory'] attribute.",
+        "Valid values are non-zero integers followed by m (e.g., 49416564kB)."
```
Comment (legal90) [inline recipes/autotune.rb]: `followed by m` should be replaced with `followed by kB`

**Comment by mvdkleijn** [comment]
@legal90 This feature is based on the config_pgtune recipe feature of the postgresql cookbook. As such I've copied over `binaryround` which is a function to truncate the given value to 4 significant bits, and render in a human readable way suitable for JVM config.

I believe its useful though it can be cleaned up a little. Which I'll do.

File: recipes/autotune.rb (hunk 2 above)
Comment (mvdkleijn) [inline recipes/autotune.rb]: Good catch :smile:

**Pull request merged**